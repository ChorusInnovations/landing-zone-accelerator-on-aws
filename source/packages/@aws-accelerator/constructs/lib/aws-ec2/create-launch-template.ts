/**
 *  Copyright 2022 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

import * as cdk from 'aws-cdk-lib';
import { Construct } from 'constructs';
import * as fs from 'fs';

export interface ILaunchTemplateResource extends cdk.IResource {
  /**
   * The version number.
   */
  readonly version: string;
  /**
   * The ID of the AWS::EC2::LaunchTemplate . You must specify either a LaunchTemplateName or a LaunchTemplateId .
   */
  readonly launchTemplateId: string;
  /**
   * The name of the AWS::EC2::LaunchTemplate . You must specify either a LaunchTemplateName or a LaunchTemplateId .
   */
  readonly launchTemplateName: string;
}
export type PrivateIpAddressConfig = {
  primary: boolean | undefined;
  privateIpAddress: string | undefined;
};
export type NetworkInterfaceItemConfig = {
  associateCarrierIpAddress: boolean | undefined;
  associatePublicIpAddress: boolean | undefined;
  deleteOnTermination: boolean | undefined;
  description: string | undefined;
  deviceIndex: number | undefined;
  groups: string[] | undefined;
  interfaceType: string | undefined;
  networkCardIndex: number | undefined;
  networkInterfaceId: string | undefined;
  privateIpAddress: string | undefined;
  secondaryPrivateIpAddressCount: number | undefined;
  subnetId: string | undefined;
  privateIpAddresses: PrivateIpAddressConfig[] | undefined;
};
export type EbsProperty = {
  deleteOnTermination?: boolean;
  encrypted?: boolean;
  iops?: number;
  kmsKeyId?: string;
  snapshotId?: string;
  throughput?: number;
  volumeSize?: number;
  volumeType?: string;
};

export type BlockDeviceMappingItem = {
  deviceName: string;
  ebs?: EbsProperty;
};

export interface LaunchTemplateProps {
  /*
   * Path to user data.
   */
  readonly userData?: string;
  /*
   * VpcName
   */
  readonly vpc: string;
  /*
   * Name of Launch Template
   */
  readonly name: string;
  /*
   * Name of Application
   */
  readonly appName: string;
  readonly blockDeviceMappings?: BlockDeviceMappingItem[];
  readonly securityGroups?: string[];
  readonly networkInterfaces?: NetworkInterfaceItemConfig[];
  readonly instanceType: string;
  readonly keyPair?: string;
  readonly iamInstanceProfile?: string;
  readonly imageId: string;
  readonly enforceImdsv2?: boolean;
}

export class LaunchTemplate extends cdk.Resource implements ILaunchTemplateResource {
  public readonly version: string;
  public readonly launchTemplateId: string;
  public readonly launchTemplateName: string;

  constructor(scope: Construct, id: string, props: LaunchTemplateProps) {
    super(scope, id);

    let metadataOptions: cdk.aws_ec2.CfnLaunchTemplate.MetadataOptionsProperty;

    if (props.enforceImdsv2 === false) {
      metadataOptions = { httpTokens: 'optional' };
    } else {
      metadataOptions = { httpTokens: 'required' };
    }

    const launchTemplate = new cdk.aws_ec2.CfnLaunchTemplate(this, 'LaunchTemplate', {
      launchTemplateData: {
        blockDeviceMappings: props.blockDeviceMappings ?? undefined,
        securityGroupIds: props.securityGroups ?? undefined,
        networkInterfaces: props.networkInterfaces ?? undefined,
        instanceType: props.instanceType,
        keyName: props.keyPair ?? undefined,
        imageId: props.imageId,
        iamInstanceProfile: { name: props.iamInstanceProfile ?? undefined },
        userData: cdk.Fn.base64(fs.readFileSync(props.userData!, 'utf8')) ?? undefined,
        metadataOptions,
      },
      launchTemplateName: props.name,
    });

    this.launchTemplateId = launchTemplate.ref;
    this.launchTemplateName = launchTemplate.launchTemplateName!;
    this.version = launchTemplate.attrLatestVersionNumber;
  }
}
