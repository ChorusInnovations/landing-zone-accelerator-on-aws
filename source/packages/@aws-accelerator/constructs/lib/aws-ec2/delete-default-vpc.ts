/**
 *  Copyright 2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

import * as cdk from 'aws-cdk-lib';
import { v4 as uuidv4 } from 'uuid';
import { Construct } from 'constructs';

const path = require('path');

/**
 * Class Delete the Default VPC
 */
export class DeleteDefaultVpc extends Construct {
  readonly id: string;
  constructor(scope: Construct, id: string) {
    super(scope, id);

    const DELETE_DEFAULT_VPC_TYPE = 'Custom::DeleteDefaultVpc';

    //
    // Function definition for the custom resource
    //
    const cr = cdk.CustomResourceProvider.getOrCreateProvider(this, DELETE_DEFAULT_VPC_TYPE, {
      codeDirectory: path.join(__dirname, 'delete-default-vpc/dist'),
      runtime: cdk.CustomResourceProviderRuntime.NODEJS_14_X,
      policyStatements: [
        {
          Effect: 'Allow',
          Action: [
            'ec2:DeleteInternetGateway',
            'ec2:DetachInternetGateway',
            'ec2:DeleteNetworkAcl',
            'ec2:DeleteRoute',
            'ec2:DeleteSecurityGroup',
            'ec2:DeleteSubnet',
            'ec2:DeleteVpc',
            'ec2:DescribeInternetGateways',
            'ec2:DescribeNetworkAcls',
            'ec2:DescribeRouteTables',
            'ec2:DescribeSecurityGroups',
            'ec2:DescribeSubnets',
            'ec2:DescribeVpcs',
          ],
          Resource: '*',
        },
      ],
    });

    //
    // Custom Resource definition. We want this resource to be evaluated on
    // every CloudFormation update, so we generate a new uuid to force
    // re-evaluation.
    //
    const resource = new cdk.CustomResource(this, 'Resource', {
      resourceType: DELETE_DEFAULT_VPC_TYPE,
      serviceToken: cr.serviceToken,
      properties: {
        uuid: uuidv4(),
      },
    });

    this.id = resource.ref;
  }
}
