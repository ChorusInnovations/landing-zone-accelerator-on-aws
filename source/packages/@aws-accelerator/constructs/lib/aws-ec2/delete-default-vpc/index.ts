/**
 *  Copyright 2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

import { throttlingBackOff } from '@aws-accelerator/utils';
import * as console from 'console';
import {
  DeleteInternetGatewayCommand,
  DetachInternetGatewayCommand,
  DeleteNetworkAclCommand,
  DeleteRouteCommand,
  DeleteSecurityGroupCommand,
  DeleteSubnetCommand,
  DeleteVpcCommand,
  EC2Client,
  paginateDescribeInternetGateways,
  paginateDescribeNetworkAcls,
  paginateDescribeRouteTables,
  paginateDescribeSecurityGroups,
  paginateDescribeSubnets,
  paginateDescribeVpcs,
} from '@aws-sdk/client-ec2';

/**
 * delete-default-vpc - lambda handler
 *
 * @param event
 * @returns
 */

export async function handler(event: AWSLambda.CloudFormationCustomResourceEvent): Promise<
  | {
    PhysicalResourceId: string | undefined;
    Status: string;
  }
  | undefined
> {
  // Retrieve operating region that stack is ran
  const region = event.ResourceProperties['region'];
  const ec2Client = new EC2Client({ region: region });
  const defaultVpcIds: string[] = [];

  switch (event.RequestType) {
    case 'Create':
    case 'Update':
      console.log(`Starting - Deletion of default VPC and associated resources in ${region}`);

      // Retrieve default VPC(s)
      for await (const page of paginateDescribeVpcs(
        { client: ec2Client },
        { Filters: [{ Name: 'is-default', Values: [`true`] }] },
      )) {
        for (const vpc of page.Vpcs ?? []) {
          if (vpc.VpcId) {
            defaultVpcIds.push(vpc.VpcId);
          }
        }
      }
      console.log(`List of VPCs: `, defaultVpcIds);
      if (defaultVpcIds.length == 0) {
        console.warn('No default VPCs detected');
        return {
          PhysicalResourceId: `delete-default-vpc`,
          Status: 'SUCCESS',
        };
      } else {
        console.warn('Default VPC Detected');
      }

      // Retrieve and detach, delete IGWs
      for (const vpcId of defaultVpcIds) {
        for await (const page of paginateDescribeInternetGateways(
          { client: ec2Client },
          { Filters: [{ Name: 'attachment.vpc-id', Values: [vpcId] }] },
        )) {
          for (const igw of page.InternetGateways ?? []) {
            for (const attachment of igw.Attachments ?? []) {
              if (attachment.State == 'available') {
                console.log(`Detaching ${igw.InternetGatewayId}`);
                await throttlingBackOff(() =>
                  ec2Client.send(
                    new DetachInternetGatewayCommand({ InternetGatewayId: igw.InternetGatewayId, VpcId: vpcId }),
                  ),
                );
              }
              console.warn(`${igw.InternetGatewayId} is not attached. Proceeding to delete.`);
              await throttlingBackOff(() =>
                ec2Client.send(
                  new DeleteInternetGatewayCommand({
                    InternetGatewayId: igw.InternetGatewayId,
                  }),
                ),
              );
            }
          }
        }

        // Retrieve Default VPC Subnets
        console.log(`Gathering Subnets for VPC ${vpcId}`);
        for await (const page of paginateDescribeSubnets(
          { client: ec2Client },
          { Filters: [{ Name: 'vpc-id', Values: [vpcId] }] },
        )) {
          for (const subnet of page.Subnets ?? []) {
            console.log(`Delete Subnet ${subnet.SubnetId}`);
            await throttlingBackOff(() =>
              ec2Client.send(
                new DeleteSubnetCommand({
                  SubnetId: subnet.SubnetId,
                }),
              ),
            );
          }
        }

        // Delete Routes
        console.log(`Gathering list of Route Tables for VPC ${vpcId}`);
        for await (const page of paginateDescribeRouteTables(
          { client: ec2Client },
          { Filters: [{ Name: 'vpc-id', Values: [vpcId] }] },
        )) {
          for (const routeTableObject of page.RouteTables ?? []) {
            for (const routes of routeTableObject.Routes ?? []) {
              if (routes.GatewayId !== 'local') {
                console.log(`Removing route ${routes.DestinationCidrBlock} from ${routeTableObject.RouteTableId}`);
                await throttlingBackOff(() =>
                  ec2Client.send(
                    new DeleteRouteCommand({
                      RouteTableId: routeTableObject.RouteTableId,
                      DestinationCidrBlock: routes.DestinationCidrBlock,
                    }),
                  ),
                );
              }
            }
          }
        }

        // List and Delete NACLs
        console.log(`Gathering list of NACLs for VPC ${vpcId}`);
        for await (const page of paginateDescribeNetworkAcls(
          { client: ec2Client },
          { Filters: [{ Name: 'vpc-id', Values: [vpcId] }] },
        )) {
          for (const networkAclObject of page.NetworkAcls ?? []) {
            if (networkAclObject.IsDefault !== true) {
              console.log(`Deleting Network ACL ID ${networkAclObject.NetworkAclId}`);
              await throttlingBackOff(() =>
                ec2Client.send(
                  new DeleteNetworkAclCommand({
                    NetworkAclId: networkAclObject.NetworkAclId,
                  }),
                ),
              );
            } else {
              console.warn(`${networkAclObject.NetworkAclId} is the default NACL. Ignoring`);
            }
          }
        }

        // List and Delete Security Groups
        console.log(`Gathering list of Security Groups for VPC ${vpcId}`);
        for await (const page of paginateDescribeSecurityGroups(
          { client: ec2Client },
          { Filters: [{ Name: 'vpc-id', Values: [vpcId] }] },
        )) {
          for (const securityGroupObject of page.SecurityGroups ?? []) {
            if (securityGroupObject.GroupName == 'default') {
              console.warn(`${securityGroupObject.GroupId} is the default SG. Ignoring`);
            } else {
              console.log(`Deleting Security Group Id ${securityGroupObject.GroupId}`);
              await throttlingBackOff(() =>
                ec2Client.send(
                  new DeleteSecurityGroupCommand({
                    GroupId: securityGroupObject.GroupId,
                  }),
                ),
              );
            }
          }
        }
        // Once all resources are deleted, delete the VPC.
        console.log(`Deleting VPC ${vpcId}`);
        await throttlingBackOff(() => ec2Client.send(new DeleteVpcCommand({ VpcId: vpcId })));
      }

      return {
        PhysicalResourceId: `delete-default-vpc`,
        Status: 'SUCCESS',
      };

    case 'Delete':
      // Do Nothing
      return {
        PhysicalResourceId: event.PhysicalResourceId,
        Status: 'SUCCESS',
      };
  }
}
