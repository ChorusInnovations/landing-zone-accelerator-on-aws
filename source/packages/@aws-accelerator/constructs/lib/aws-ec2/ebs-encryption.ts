/**
 *  Copyright 2022 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

import * as cdk from 'aws-cdk-lib';
import { Construct } from 'constructs';

const path = require('path');

/**
 * Initialized EbsVolumeEncryptionProps properties
 */
export interface EbsVolumeEncryptionProps {
  readonly kmsKey: cdk.aws_kms.IKey;
}

/**
/**
* Class to Enable Default EBS Volume Encryption
*/
export class EbsDefaultEncryption extends Construct {
  public readonly id: string;

  constructor(scope: Construct, id: string, props: EbsVolumeEncryptionProps) {
    super(scope, id);

    const EBS_ENCRYPTION_TYPE = 'Custom::EbsDefaultVolumeEncryption';

    const iamPolicy = [
      {
        Sid: 'EC2',
        Effect: 'Allow',
        Action: [
          'ec2:DisableEbsEncryptionByDefault',
          'ec2:EnableEbsEncryptionByDefault',
          'ec2:ModifyEbsDefaultKmsKeyId',
          'ec2:ResetEbsDefaultKmsKeyId',
        ],
        Resource: '*',
      },
      {
        Sid: 'KMS',
        Effect: 'Allow',
        Action: ['kms:DescribeKey'],
        Resource: props.kmsKey.keyArn,
      },
    ];

    const enableEncryptionFunction = cdk.CustomResourceProvider.getOrCreateProvider(this, EBS_ENCRYPTION_TYPE, {
      codeDirectory: path.join(__dirname, 'ebs-default-encryption/dist'),
      runtime: cdk.CustomResourceProviderRuntime.NODEJS_14_X,
      policyStatements: iamPolicy,
    });

    const resource = new cdk.CustomResource(this, 'Resource', {
      resourceType: EBS_ENCRYPTION_TYPE,
      serviceToken: enableEncryptionFunction.serviceToken,
      properties: {
        kmsKeyId: props.kmsKey?.keyId,
      },
    });

    this.id = resource.ref;
  }
}
