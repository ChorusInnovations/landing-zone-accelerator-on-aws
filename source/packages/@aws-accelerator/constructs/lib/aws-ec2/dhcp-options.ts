/**
 *  Copyright 2022 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

import { CfnTag, IResource, Resource } from 'aws-cdk-lib';
import * as ec2 from 'aws-cdk-lib/aws-ec2';
import { Construct } from 'constructs';

export interface IDhcpOptions extends IResource {
  /**
   * The name of the DHCP options set.
   */
  readonly name: string;

  /**
   * The ID of the DHCP options set.
   */
  readonly dhcpOptionsId: string;
}

export interface DhcpOptionsProps {
  /**
   * The name of the DHCP options set.
   */
  readonly name: string;

  /**
   * This value is used to complete unqualified DNS hostnames.
   */
  readonly domainName?: string;

  /**
   * The IPv4 addresses of up to four domain name servers.
   *
   * @default -- AmazonProvidedDNS
   */
  readonly domainNameServers?: string[];

  /**
   * The IPv4 addresses of up to four NetBIOS name servers.
   */
  readonly netbiosNameServers?: string[];

  /**
   * The NetBIOS node type (1, 2, 4, or 8).
   */
  readonly netbiosNodeType?: number;

  /**
   * The IPv4 addresses of up to four Network Time Protocol (NTP) servers.
   */
  readonly ntpServers?: string[];

  /**
   * Any tags assigned to the DHCP options set.
   */
  readonly tags?: CfnTag[];
}

export class DhcpOptions extends Resource implements IDhcpOptions {
  public readonly name: string;
  public readonly dhcpOptionsId: string;

  constructor(scope: Construct, id: string, props: DhcpOptionsProps) {
    super(scope, id);

    this.name = props.name;

    // Add name tag to tags
    props.tags?.push({ key: 'Name', value: this.name });

    const resource = new ec2.CfnDHCPOptions(this, 'Resource', {
      domainName: props.domainName,
      domainNameServers: props.domainNameServers ?? ['AmazonProvidedDNS'],
      netbiosNameServers: props.netbiosNameServers,
      netbiosNodeType: props.netbiosNodeType,
      ntpServers: props.ntpServers,
      tags: props.tags,
    });

    this.dhcpOptionsId = resource.attrDhcpOptionsId;
  }
}
