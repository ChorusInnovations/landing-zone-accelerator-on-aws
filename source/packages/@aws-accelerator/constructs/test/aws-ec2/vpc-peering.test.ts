/**
 *  Copyright 2022 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

import * as cdk from 'aws-cdk-lib';

import { VpcPeering } from '../../lib/aws-ec2/vpc-peering';

const testNamePrefix = 'Construct(VpcPeering): ';

//Initialize stack for tests
const stack = new cdk.Stack();

new VpcPeering(stack, 'TestPeering', {
  name: 'Test',
  peerOwnerId: '111111111111',
  peerRegion: 'us-east-1',
  peerVpcId: 'AccepterVpc',
  vpcId: 'RequesterVpc',
  peerRoleName: 'TestRole',
  tags: [],
});

/**
 * VPC peering construct test
 */
describe('VpcPeering', () => {
  /**
   * Number of VPC peering test
   */
  test(`${testNamePrefix} VPC peering count test`, () => {
    cdk.assertions.Template.fromStack(stack).resourceCountIs('AWS::EC2::VPCPeeringConnection', 1);
  });

  /**
   * VPC peering resource configuration test
   */
  test(`${testNamePrefix} VPC peering resource configuration test`, () => {
    cdk.assertions.Template.fromStack(stack).templateMatches({
      Resources: {
        TestPeeringF63C5812: {
          Type: 'AWS::EC2::VPCPeeringConnection',
          Properties: {
            PeerOwnerId: '111111111111',
            PeerRegion: 'us-east-1',
            PeerRoleArn: {
              'Fn::Join': ['', ['arn:', { Ref: 'AWS::Partition' }, ':iam::111111111111:role/TestRole']],
            },
            PeerVpcId: 'AccepterVpc',
            Tags: [
              {
                Key: 'Name',
                Value: 'Test',
              },
            ],
            VpcId: 'RequesterVpc',
          },
        },
      },
    });
  });
});
