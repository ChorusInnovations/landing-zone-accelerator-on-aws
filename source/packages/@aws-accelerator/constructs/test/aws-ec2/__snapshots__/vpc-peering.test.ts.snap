// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VpcPeering Construct(VpcPeering):  Snapshot Test 1`] = `
Object {
  "Resources": Object {
    "CustomPrefixListRouteCustomResourceProviderHandler5B28D077": Object {
      "DependsOn": Array [
        "CustomPrefixListRouteCustomResourceProviderRoleD08268B5",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "REPLACED-GENERATED-NAME.zip",
        },
        "Handler": "__entrypoint__.handler",
        "MemorySize": 128,
        "Role": Object {
          "Fn::GetAtt": Array [
            "CustomPrefixListRouteCustomResourceProviderRoleD08268B5",
            "Arn",
          ],
        },
        "Runtime": "nodejs14.x",
        "Timeout": 900,
      },
      "Type": "AWS::Lambda::Function",
    },
    "CustomPrefixListRouteCustomResourceProviderLogGroup68DB81A5": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "KmsKeyId": Object {
          "Fn::GetAtt": Array [
            "kmsKey2EF55D91A",
            "Arn",
          ],
        },
        "LogGroupName": Object {
          "Fn::Join": Array [
            "",
            Array [
              "/aws/lambda/",
              Object {
                "Ref": "CustomPrefixListRouteCustomResourceProviderHandler5B28D077",
              },
            ],
          ],
        },
        "RetentionInDays": 10,
      },
      "Type": "AWS::Logs::LogGroup",
      "UpdateReplacePolicy": "Delete",
    },
    "CustomPrefixListRouteCustomResourceProviderRoleD08268B5": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Sub": "arn:\${AWS::Partition}:iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
          },
        ],
        "Policies": Array [
          Object {
            "PolicyDocument": Object {
              "Statement": Array [
                Object {
                  "Action": Array [
                    "ec2:CreateRoute",
                    "ec2:DeleteRoute",
                  ],
                  "Effect": "Allow",
                  "Resource": "*",
                  "Sid": "AllowModifyRoutes",
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "Inline",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "TestPeeringF63C5812": Object {
      "Properties": Object {
        "PeerOwnerId": "111111111111",
        "PeerRegion": "us-east-1",
        "PeerRoleArn": Object {
          "Fn::Join": Array [
            "",
            Array [
              "arn:",
              Object {
                "Ref": "AWS::Partition",
              },
              ":iam::111111111111:role/TestRole",
            ],
          ],
        },
        "PeerVpcId": "AccepterVpc",
        "Tags": Array [
          Object {
            "Key": "key",
            "Value": "value",
          },
          Object {
            "Key": "Name",
            "Value": "Test",
          },
        ],
        "VpcId": "RequesterVpc",
      },
      "Type": "AWS::EC2::VPCPeeringConnection",
    },
    "TestPeeringcrossAccountPeerRouteA647E009": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "ServiceToken": Object {
          "Fn::GetAtt": Array [
            "myProviderframeworkonEventAB595BBA",
            "Arn",
          ],
        },
        "region": Object {
          "Ref": "AWS::Region",
        },
        "roleArn": Object {
          "Fn::Join": Array [
            "",
            Array [
              "arn:",
              Object {
                "Ref": "AWS::Partition",
              },
              ":iam::111111111111:role/role",
            ],
          ],
        },
        "routeDefinition": Object {
          "DestinationCidrBlock": "10.0.0.6/32",
          "RouteTableId": "rt-3456",
          "VpcPeeringConnectionId": Object {
            "Ref": "TestPeeringF63C5812",
          },
        },
      },
      "Type": "Custom::CrossAccountRoute",
      "UpdateReplacePolicy": "Delete",
    },
    "TestPeeringvpcPeeringRoute112B3732": Object {
      "Properties": Object {
        "DestinationCidrBlock": "10.0.0.5/32",
        "RouteTableId": "rt-12345",
        "VpcPeeringConnectionId": Object {
          "Ref": "TestPeeringF63C5812",
        },
      },
      "Type": "AWS::EC2::Route",
    },
    "TestPeeringvpcPeeringRoute313B0AACD": Object {
      "DeletionPolicy": "Delete",
      "DependsOn": Array [
        "CustomPrefixListRouteCustomResourceProviderLogGroup68DB81A5",
      ],
      "Properties": Object {
        "ServiceToken": Object {
          "Fn::GetAtt": Array [
            "CustomPrefixListRouteCustomResourceProviderHandler5B28D077",
            "Arn",
          ],
        },
        "routeDefinition": Object {
          "DestinationPrefixListId": "destinationRoute",
          "RouteTableId": "rt-12345",
          "VpcPeeringConnectionId": Object {
            "Ref": "TestPeeringF63C5812",
          },
        },
      },
      "Type": "Custom::PrefixListRoute",
      "UpdateReplacePolicy": "Delete",
    },
    "kmsKey1A3FA0EFA": Object {
      "DeletionPolicy": "Retain",
      "Properties": Object {
        "KeyPolicy": Object {
          "Statement": Array [
            Object {
              "Action": "kms:*",
              "Effect": "Allow",
              "Principal": Object {
                "AWS": Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":iam::",
                      Object {
                        "Ref": "AWS::AccountId",
                      },
                      ":root",
                    ],
                  ],
                },
              },
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::KMS::Key",
      "UpdateReplacePolicy": "Retain",
    },
    "kmsKey2EF55D91A": Object {
      "DeletionPolicy": "Retain",
      "Properties": Object {
        "KeyPolicy": Object {
          "Statement": Array [
            Object {
              "Action": "kms:*",
              "Effect": "Allow",
              "Principal": Object {
                "AWS": Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":iam::",
                      Object {
                        "Ref": "AWS::AccountId",
                      },
                      ":root",
                    ],
                  ],
                },
              },
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::KMS::Key",
      "UpdateReplacePolicy": "Retain",
    },
    "kmsKey3992EC0B2": Object {
      "DeletionPolicy": "Retain",
      "Properties": Object {
        "KeyPolicy": Object {
          "Statement": Array [
            Object {
              "Action": "kms:*",
              "Effect": "Allow",
              "Principal": Object {
                "AWS": Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":iam::",
                      Object {
                        "Ref": "AWS::AccountId",
                      },
                      ":root",
                    ],
                  ],
                },
              },
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::KMS::Key",
      "UpdateReplacePolicy": "Retain",
    },
    "kmsKey3BB021CC": Object {
      "DeletionPolicy": "Retain",
      "Properties": Object {
        "KeyPolicy": Object {
          "Statement": Array [
            Object {
              "Action": "kms:*",
              "Effect": "Allow",
              "Principal": Object {
                "AWS": Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":iam::",
                      Object {
                        "Ref": "AWS::AccountId",
                      },
                      ":root",
                    ],
                  ],
                },
              },
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::KMS::Key",
      "UpdateReplacePolicy": "Retain",
    },
    "myProviderframeworkonEventAB595BBA": Object {
      "DependsOn": Array [
        "myProviderframeworkonEventServiceRoleDefaultPolicy4EAC636C",
        "myProviderframeworkonEventServiceRole94BD41C2",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "REPLACED-GENERATED-NAME.zip",
        },
        "Description": "AWS CDK resource provider framework - onEvent (Default/myProvider)",
        "Environment": Object {
          "Variables": Object {
            "USER_ON_EVENT_FUNCTION_ARN": Object {
              "Fn::GetAtt": Array [
                "testAF53AC38",
                "Arn",
              ],
            },
          },
        },
        "Handler": "framework.onEvent",
        "Role": Object {
          "Fn::GetAtt": Array [
            "myProviderframeworkonEventServiceRole94BD41C2",
            "Arn",
          ],
        },
        "Runtime": "nodejs14.x",
        "Timeout": 900,
      },
      "Type": "AWS::Lambda::Function",
    },
    "myProviderframeworkonEventServiceRole94BD41C2": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "myProviderframeworkonEventServiceRoleDefaultPolicy4EAC636C": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "lambda:InvokeFunction",
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::GetAtt": Array [
                    "testAF53AC38",
                    "Arn",
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      Object {
                        "Fn::GetAtt": Array [
                          "testAF53AC38",
                          "Arn",
                        ],
                      },
                      ":*",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "myProviderframeworkonEventServiceRoleDefaultPolicy4EAC636C",
        "Roles": Array [
          Object {
            "Ref": "myProviderframeworkonEventServiceRole94BD41C2",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "testAF53AC38": Object {
      "DependsOn": Array [
        "testServiceRole78466FFF",
      ],
      "Properties": Object {
        "Code": Object {
          "ZipFile": "foo",
        },
        "Handler": "handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "testServiceRole78466FFF",
            "Arn",
          ],
        },
        "Runtime": "nodejs14.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "testServiceRole78466FFF": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
  },
}
`;
